src/braket/pennylane_plugin/ahs_device.py

```diff
@@ -0,0 +1,515 @@
+# language governing permissions and limitations under the License.
+
+"""
+Devices
+=======
+
+**Module name:** :mod:`braket.pennylane_braket.ahs_device`
+
+.. currentmodule:: braket.pennylane_braket.ahs_device
+
+Braket Analogue Hamiltonian Simulation (AHS) devices to be used with PennyLane
+
+Classes
+-------
+
+.. autosummary::
+   BraketAwsAhsDevice
+   BraketLocalAhsDevice
+
+Code details
+~~~~~~~~~~~~
+"""
+from functools import partial
+from typing import Iterable, Union, Optional
+import numpy as np
+
+from pennylane import QubitDevice
+from pennylane._version import __version__
+from pennylane.pulse.hardware_hamiltonian import HardwarePulse, HardwareHamiltonian
+
+from braket.aws import AwsDevice, AwsSession, AwsQuantumTask
+from braket.devices import Device, LocalSimulator
+from braket.ahs.atom_arrangement import AtomArrangement
+from braket.ahs.analog_hamiltonian_simulation import AnalogHamiltonianSimulation
+from braket.ahs.driving_field import DrivingField
+from braket.timings.time_series import TimeSeries
+
+
+class BraketAhsDevice(QubitDevice):
+    """Abstract Amazon Braket device for analogue hamiltonian simulation with PennyLane.
+
+    Args:
+        wires (int or Iterable[Number, str]]): Number of subsystems represented by the device,
+            or iterable that contains unique labels for the subsystems as numbers
+            (i.e., ``[-1, 0, 2]``) or strings (``['ancilla', 'q1', 'q2']``).
+        device (Device): The Amazon Braket device to use with PennyLane.
+        shots (int): Number of executions to run to aquire measurements. Defaults to 100.
+    """
+
+    name = "Braket AHS PennyLane plugin"
+    pennylane_requires = ">=0.30.0"
+    version = __version__
+    author = "Xanadu Inc."
+
+    operations = {"ParametrizedEvolution"}
+
+    def __init__(
+            self,
+            wires: Union[int, Iterable],
+            device: Device, *,
+            shots: int = AwsDevice.DEFAULT_SHOTS_QPU
+    ):
+        if not shots:
+            raise RuntimeError(f"This device requires shots. Received shots={shots}")
+        self._device = device
+
+        super().__init__(wires=wires, shots=shots)
+
+        self.register = None
+        self.pulses = None
+        self.ahs_program = None
+        self.samples = None
+
+    def apply(self, operations, **kwargs):
+        """Convert the pulse operation to an AHS program and run on the connected device"""
+
+        if not np.all([op.name in self.operations for op in operations]):
+            raise NotImplementedError(
+                "Device {self.short_name} expected only operations "
+                "{self.operations} but recieved {operations}"
+            )
+
+        self._validate_operations(operations)
+
+        ev_op = operations[0]  # only one!
+
+        self._validate_pulses(ev_op.H.pulses)
+
+        ahs_program = self.create_ahs_program(ev_op)
+
+        task = self._run_task(ahs_program)
+
+        self.samples = task.result()
```

Comment:
Formatting nit: easy enough to read without the extra spaces
```suggestion
        self._validate_operations(operations)
        ev_op = operations[0]  # only one!

        self._validate_pulses(ev_op.H.pulses)
        ahs_program = self.create_ahs_program(ev_op)
        task = self._run_task(ahs_program)
        self.samples = task.result()
```